File: ex5.hs
-- -- Exercise 5
-- -- search for the local maximum of f nearest x using an
-- -- approximation margin delta and initial step value s
-- goldenRatio = (1+ sqrt 5) / 2
-- invphi = (sqrt(5) - 1) / 2 -- 1/phi
-- invphi2 = (3 - sqrt(5)) / 2 -- 1/phi^2



-- -- TODO: the incorrect part is in the transfering argument of 2 functions

-- tol=1e-5
-- test = gssRecursive (\x -> 4.0-x*x) (-2.5) 2.5 tol 0 0 0 0 0
-- test3 = hillClimb (\x -> 4.0-x*x) (-2.5) 2.5 1e-3

goldenRatio = (1 + sqrt 5)/2
hillClimb :: (Float -> Float) -> Float -> Float -> Float -> Float
hillClimb f x y tol | (b - a) <= tol = a
                    | (f c) >= (f d) = hillClimb f a d tol
                    | (f c) < (f d) = hillClimb f c b tol   
 where 
  a = min x y
  b = max x y
  c = a + (b-a)/(goldenRatio+1)
  d = b - (b-a)/(goldenRatio+1)

  

File: ex5.py
goldenRatio = (1+math.sqrt(5))/2
invphi = (math.sqrt(5) - 1) / 2 # 1/phi                                                                                                                     
invphi2 = (3 - math.sqrt(5)) / 2 # 1/phi^2                                                                                                                  

def gssrec(f,a,b,tol=1e-5,h=None,c=None,d=None,fc=None,fd=None):
    '''                                                                                                                                                     
    Golden section search, recursive.                                                                                                                           
                                                                                                                                                            
    Given a function f with a single local minimum in                                                                                                       
    the interval [a,b], gss returns a subset interval                                                                                                       
    [c,d] that contains the minimum with d-c <= tol.                                                                                                        
                                                                                                                                                            
    example:                                                                                                                                                
    >>> f = lambda x: (x-2)**2                                                                                                                              
    >>> a = 1                                                                                                                                               
    >>> b = 5                                                                                                                                               
    >>> tol = 1e-5                                                                                                                                          
    >>> (c,d) = gssrec(f, a, b, tol)                                                                                                                           
    >>> print (c,d)                                                                                                                                         
    (1.9999959837979107, 2.0000050911830893)                                                                                                                
    '''
    
    (a,b)=(min(a,b),max(a,b))
    if h == None: h=b-a
    if h <= tol: return (a,b)
    if c == None: c = a + invphi2*h
    if d == None: d = a + invphi*h
    if fc == None: fc = f(c)
    if fd == None: fd = f(d)
    if fc < fd:
        return gssrec(f,a,d,tol,h*invphi,c=None,fc=None,d=c,fd=fc)
    else:
        return gssrec(f,c,b,tol,h*invphi,c=d,fc=fd,d=None,fd=None)

def test5():